アクセストークン無効化用のアクションを追加

a5ob7r opened the pull request:
# 概要

アクセストークンを失効させるためのアクションを追加する。

また、APIドキュメントに記載していたtoken_type_hintパラメータを削除した。本APIはaccess_tokenしかサポートしていないので、このパラメータを積極的にサポートする理由はない。

a5ob7r committed:
Add the action to revoke access tokens

a5ob7r committed:
Remove token_type_hint from supported parameters

This API only supports access_token, so we no need to support this
parameter. And this parameter doesn't have any effect whether the
request succeeds or not, only effects for token searching.

tk0miya commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -1,8 +1,8 @@
 class AccessTokensController < ApplicationController
-  before_action :validate_required_params
-  before_action :validate_grant_type
+  before_action :validate_required_params, only: :post
+  before_action :validate_grant_type, only: :post
```
他の PR でアクション名を変更していたのでご注意をー

a5ob7r replied:
おっと、修正します。

KT-tim commented on app/controllers/access_tokens_controller.rb:
```diff
@@ -23,6 +23,15 @@ def create
+  def destroy
+    access_token = AccessToken.find_by token: params[:token]
+
+    render problem: { error: 'invalid_request' }, status: :bad_request and return unless access_token
```
見つからなかった場合ならnot foundの方が適切かと思ったんですが、リクエストが変なのは確かなので、間違ったことを言っているわけではないんですよね　うーん

a5ob7r replied:
この仕様は、アクセストークン無効化に関するRFCに従ったものです。

https://datatracker.ietf.org/doc/html/rfc7009#section-2.2.1

実際には、アクセストークン発行時のエラーレスポンスにプラスアルファしたものです。

> The error presentation conforms to the definition in Section 5.2 of
   [RFC6749].  The following additional error code is defined for the
   token revocation endpoint:

そして、アクセストークン発行時のエラーレスポンスでは、基本的に400を返すとあります。今回もこれに該当すると考えます。

> The authorization server responds with an HTTP 400 (Bad Request)
   status code (unless specified otherwise) and includes the following
   parameters with the response:

https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

a5ob7r replied:
というか、不正なアクセストークンの場合でも、200を返すことになっているのを忘れていました。

> Note: invalid tokens do not cause an error response since the client
   cannot handle such an error in a reasonable way.  Moreover, the
   purpose of the revocation request, invalidating the particular token,
   is already achieved.

https://datatracker.ietf.org/doc/html/rfc7009#section-2.2

なので、今回の場合であれば、エラーレスポンスを返すのは、アクセストークンがパラメータに含まれていない場合のみになります。

a5ob7r replied:
いや、これは不正なアクセストークンではなく、すでに無効化されているトークンという意味かな。

KT-tim replied:
400を返すのが正式な仕様だったんですね　なるほど

a5ob7r commented:
今気がつきましたが、トークンの無効化はDELETEメソッドではなくPOSTメソッドでやる必要がありました。

> The client requests the revocation of a particular token by making an
   HTTP POST request to the token revocation endpoint URL.  This URL
   MUST conform to the rules given in [[RFC6749], Section 3.1](https://datatracker.ietf.org/doc/html/rfc6749#section-3.1).  Clients
   MUST verify that the URL is an HTTPS URL.

https://datatracker.ietf.org/doc/html/rfc7009#section-2

ということで修正します。

a5ob7r committed:
Change the revocation request method from DELETE to POST

This is according to the RFC.

https://datatracker.ietf.org/doc/html/rfc7009#section-2

a5ob7r committed:
Add a missing test case for a revoked token revocation request

a5ob7r committed:
Fix some test descriptions to make them more details

a5ob7r committed:
Change a test string generation method

Faker::String.random sometimes generates null contained string, which
cause an error.

a5ob7r commented:
修正が完了しました。

- アクセストークン失効用のコントローラとアクションを追加(AccessTokenRevocationsController#create)
- アクセストークンがすでに無効化されている場合の無効化リクエストのテストケースを追加
